libvips/foreign/spngsave.c

```diff
@@ -583,43 +583,60 @@ vips_foreign_save_spng_build( VipsObject *object )
         if( vips_object_argument_isset( object, "colours" ) ) 
 		spng->bitdepth = ceil( log2( spng->colours ) );
 
-        if( !vips_object_argument_isset( object, "bitdepth" ) ) 
-		spng->bitdepth = in->BandFmt == VIPS_FORMAT_UCHAR ? 8 : 16;
-
-	/* Filtering usually reduces the compression ratio for palette images,
-	 * so default off.
-	 */
-        if( !vips_object_argument_isset( object, "filter" ) &&
-		spng->palette )
-		spng->filter = VIPS_FOREIGN_PNG_FILTER_NONE;
-
-	/* If this is a RGB or RGBA image and a low bit depth has been
-	 * requested, enable palettization.
+	/* If no output bitdepth has been specified, use input Type to pick.
+	 * We only go for 16 bits for the types where we know there's a
+	 * 0-65535 range.
 	 */
-        if( in->Bands > 2 &&
-		spng->bitdepth < 8 )
-		spng->palette = TRUE;
+        if( !vips_object_argument_isset( object, "bitdepth" ) ) 
+		spng->bitdepth = 
+			(in->Type == VIPS_INTERPRETATION_RGB16 ||
+			 in->Type == VIPS_INTERPRETATION_GREY16) ? 16 : 8;
 
-	/* save->ready will have been converted to uint16 for high-bitdepth
-	 * formats (eg. float) ... we need to check Type to see if we want 
-	 * to save as 8 or 16-bits. Eg. imagine a float image tagged as sRGB.
-	 */
-	if( spng->bitdepth == 16 ) 
+	if( spng->bitdepth > 8 ) 
 		interpretation = in->Bands > 2 ?
 			VIPS_INTERPRETATION_RGB16 : 
 			VIPS_INTERPRETATION_GREY16;
 	else
 		interpretation = in->Bands > 2 ?
 			VIPS_INTERPRETATION_sRGB : 
 			VIPS_INTERPRETATION_B_W;
-	
+
 	if( vips_colourspace( in, &x, interpretation, NULL ) ) {
 		g_object_unref( in );
 		return( -1 );
 	}
 	g_object_unref( in );
 	in = x;
 
+	/* colourspace will leave float srgb as float. We need to force to
+	 * uchar in these cases.
+	 */
+	if( in->Type == VIPS_INTERPRETATION_sRGB ||
+		in->Type == VIPS_INTERPRETATION_B_W ) {
+		VipsImage *x;
+
+		if( vips_cast( in, &x, VIPS_FORMAT_UCHAR, NULL ) ) {
+			g_object_unref( in );
+			return( -1 );
+		}
+		g_object_unref( in );
+		in = x;
+	}
+
+	/* Filtering usually reduces the compression ratio for palette images,
```

Comment:
spng already does this optimization https://github.com/randy408/libspng/blob/v0.7.1/spng/spng.c#L4727-L4730

Reply:
Oh, nice. Fixed, thanks.